### PR TITLE
feat(interception): Implement request.mockResponse method

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch implements a new `Request.mockResponse` method. This
allows users to fulfill the intercepted request with a hand-crafted
response if they wish so.

References #1020.

### CODE REVIEW

File: lib/NetworkManager.js [hunk 1]
```diff
@@ -327,6 +327,46 @@ class Request {
     });
   }
 
+  async mockResponse(response = {}) {
+    // DataURL's are not interceptable. In this case, do nothing.
+    if (this.url.startsWith('data:'))
```
Comment (JoelEinbinder) [inline lib/NetworkManager.js]: Should we throw here instead of failing silently?

Comment (aslushnikov) [reply lib/NetworkManager.js]: Good call.

Comment (aslushnikov) [reply lib/NetworkManager.js]: Left it as-is for the following code to work:

```js
page.on('request', request => request.respond({
  status: 404,
  body: Buffer.from('<div>Not Found!</div>')
});
```

File: lib/NetworkManager.js [hunk 2]
```diff
@@ -341,6 +341,49 @@ class Request {
...
+    const statusCode = response.status || 200;
+    const statusText = statusTexts[statusCode] || '';
+    responseLines.push(`HTTP/1.1 ${statusCode} ${statusText}`);
+    const headers = {};
+    if (response.headers) {
+      for (const header of Object.keys(response.headers))
+        headers[header.toLowerCase()] = response.headers[header];
+    }
+    if (response.body.length && !('content-length' in headers))
+      headers['content-length'] = response.body.length;
```
Comment (paambaati) [inline lib/NetworkManager.js]: @aslushnikov Do you want to perhaps use `Buffer.byteLength(body)` instead of `body.length`?

Comment (aslushnikov) [reply lib/NetworkManager.js]: @paambaati this is reasonable, thanks.

File: test/test.js [hunk 3]
```diff
@@ -1339,6 +1339,44 @@ describe('Page', function() {
...
+    it('should allow mocking binary responses', SX(async function() {
+      await page.setRequestInterceptionEnabled(true);
+      const imageBuffer = fs.readFileSync(path.join(__dirname, 'assets', 'pptr.png'));
+      page.on('request', request => {
+        request.respond({
+          contentType: 'image/png',
+          body: imageBuffer
+        });
+      });
+      await page.goto(EMPTY_PAGE);
```
Comment (JoelEinbinder) [inline test/test.js]: I am confused what this test is doing. You are loading the image, and intercepting it with the same exact image? And won't the initial navigation also be overridden as the image?

I would expect there to be two images. And then bad.png is overridden with good.png.

File: lib/NetworkManager.js [hunk 4]
```diff
@@ -341,6 +341,49 @@ class Request {
   }
 
   /**
+   * @param {!Object} response
+   */
+  async respond(response = {}) {
+    // DataURL's are not interceptable. In this case, do nothing.
```
Comment (JoelEinbinder) [inline lib/NetworkManager.js]: Let's put a note in the API.

File: docs/api.md [hunk 5]
```diff
@@ -1897,6 +1899,32 @@ Contains the request's post body, if any.
...
+
+An example of fulfilling all requests with 404 responses:
+
+```js
+await page.setRequestInterceptionEnabled(true);
+page.on('request', request => {
+  request.respond({
+    status: 404,
+    contentType: 'text/plain',
+    body: 'Not Found!'
```
Comment (JoelEinbinder) [inline docs/api.md:1927]: Body is a buffer, but the example uses a string. Should the type on body be Buffer|string?

File: docs/api.md [hunk 6]
```diff
@@ -1897,6 +1899,32 @@ Contains the request's post body, if any.
...
+#### request.respond(response)
+- `response` <[Object]> Response that will fulfill this request
+  - `status` <[number]> Response status code, defaults to `200`.
+  - `headers` <[Object]> Optional response headers
+  - `contentType` <[string]> If set, equals to setting `Content-Type` response header
+  - `body` <[Buffer]> Optional response body
+- returns: <[Promise]>
+
+Fulfills request with given response. To use this, request interception should
+be enabled with `page.setRequestInterceptionEnabled`. Exception is immediately
```
Comment (JoelEinbinder) [inline docs/api.md]: An exception is thrown if request interception is not enabled.

File: test/test.js [hunk 7]
```diff
@@ -1339,6 +1339,44 @@ describe('Page', function() {
...
+      await page.evaluate(() => {
+        const img = document.createElement('img');
+        img.src = '/pptr.png';
+        document.body.appendChild(img);
+        return new Promise(fulfill => img.onload = fulfill);
+      });
+      const img = await page.$('img');
+      expect(await img.screenshot()).toBeGolden('mock-binary-response.png');
+    }));
+  });
```
Comment (JoelEinbinder) [inline test/test.js]: Add a test to throw if interception is not enabled or if trying to respond twice to the same request?

File: lib/NetworkManager.js [hunk 8]
```diff
@@ -341,6 +341,49 @@ class Request {
   }
 
   /**
+   * @param {!Object} response
```
Comment (JoelEinbinder) [inline lib/NetworkManager.js]: I'd feel safer if the options were typed here, considering we are getting a buffer.

File: docs/api.md (hunk 6 above)
Comment (aslushnikov) [reply docs/api.md]: Done

File: docs/api.md (hunk 5 above)
Comment (aslushnikov) [reply docs/api.md:1927]: Done.

File: lib/NetworkManager.js (hunk 4 above)
Comment (aslushnikov) [reply lib/NetworkManager.js]: Done.

File: test/test.js (hunk 3 above)
Comment (aslushnikov) [reply test/test.js]: This is a funny one :) Fixed.

File: test/test.js (hunk 7 above)
Comment (aslushnikov) [reply test/test.js]: Done.

File: lib/NetworkManager.js (hunk 8 above)
Comment (aslushnikov) [reply lib/NetworkManager.js]: Done.

File: docs/api.md [hunk 9]
```diff
@@ -994,6 +996,9 @@ puppeteer.launch().then(async browser => {
 });
 ```
 
+> **NOTE** Request interception doesn't work with data URLs. Calling `abort`,
+> `continue` or `respond` on requests for data URLs is a noop.
```
Comment (JoelEinbinder) [inline docs/api.md:1000]: no-op

**Review by JoelEinbinder — APPROVED** [review]

File: docs/api.md (hunk 9 above)
Comment (aslushnikov) [reply docs/api.md:1000]: Found "noop" in google, it's apparently a thing as well :)

**Pull request merged**